### ISSUE TITLE
Document conditions for fastpath in manual?

### ISSUE THREAD

**Issue by Ivan-Velickovic** [seL4/seL4#1086]
Given that the IPC fastpath is enabled by default, I was wondering whether it is worth documenting in the manual the conditions for the fastpath to actually be taken so that users are aware how to optimise for its use?

**Comment by andybui01** [comment]
Yes, my anecdote is that I use the seL4 docs/tutorial to refresh myself on the fastpath condition, it would be easier if it was in the manual :)

**Comment by Ivan-Velickovic** [comment]
> Yes, my anecdote is that I use the seL4 docs/tutorial to refresh myself on the fastpath condition, it would be easier if it was in the manual :)

Yes I often forget all the conditions. Having a section in the manual just describing the IPC fastpath would be good. It might also be worth mentioning the signal fastpath.

**Comment by Indanz** [comment]
The signal fastpath is a configuration option that is default off I think, so it would be confusing to mention.

If implementation details like these are added, it should be a new section or appendix. It would be good to add generic information about how to use and how not to use seL4, like mentioned here:

https://microkerneldude.org/2019/03/07/how-to-and-how-not-to-use-sel4-ipc/

Fastpath conditions are I think easy: IPC buffer not used and no SC switch, only immediate task-switch.

**Comment by andybui01** [comment]
The doc lists 5 dot points for fastpath conditions and it might not be so simple for beginners:
- seL4_Call or seL4_ReplyRecv must be used.
- The data in the message must fit into the seL4_FastMessageRegisters registers.
- The processes must have valid address spaces.
- No caps should be transferred.
- No other threads in the scheduler of higher priority than the thread unblocked by the IPC can be running.

It would not hurt to spell out the exact conditions for a fastpath in the manual, and we should try shepherd beginners to common/good seL4 practice when possible.

**Comment by Indanz** [comment]
For MCS, the server should be passive too.
